File: src/1405-Longest-Happy-String/1405.py
class Solution:
    def longestDiverseString(self, A, B, C, a='a', b='b', c='c'):
        if A < B:
            return self.longestDiverseString(B, A, C, b, a, c)
        if B < C:
            return self.longestDiverseString(A, C, B, a, c, b)
        if B == 0:
            return min(2, A) * a
        na = min(2, A)
        nb = 1 if A - na >= B else 0
        return na * a + nb * b + self.longestDiverseString(A - na, B - nb, C, a, b, c)

File: src/1405-Longest-Happy-String/1405.cpp
class Solution {
public:
    string longestDiverseString(int A, int B, int C, char a='a', char b='b', char c='c') {
        if (A < B) return longestDiverseString(B, A, C, b, a, c);
        if (B < C) return longestDiverseString(A, C, B, a, c, b);
        if (B == 0) return string(min(2, A), a);
        int na = min(2, A), nb = A - na + 1 >= B ? 1 : 0;
        return string(na, a) + string(nb, b) + longestDiverseString(A - na, B - nb, C, a, b, c);
    }
};